### PR TITLE
Attempt ssh connections with and without mfa at the same time

### PR DESCRIPTION
**Pull request by rosstimothy**
`tsh ssh` would fallback to doing the mfa ceremony if connecting to the node with the already provisioned certificates failed with an access denied error. This incurs the cost of a round trip to the target host when per session mfa is required. To combat the additional latency when per session mfa is required we can attempt both the connection with the certs on hand AND start the per session mfa flow at the same time. If per session mfa is not required the client won't attempt the mfa ceremony which adds no impact there. If per session mfa is required the initial connection to the host is going to fail so the mfa ceremony will need to be performed any how.

For this to work we need to ensure that users are not prompted for mfa if completing the mfa ceremony will not actually help the user gain access to the host. If users just flat out do not have access to the host we don't want to confuse them by prompting them to touch a hardware key. Since `tsh` first calls
`proto.AuthService/IsMFARequired` before initiating the mfa ceremony we are guaranteed not to initiate the mfa ceremony when not required.

### CODE REVIEW

File: lib/client/api.go [hunk 1]
```diff
@@ -1457,78 +1457,127 @@ func (tc *TeleportClient) SSH(ctx context.Context, command []string, runLocally
...
+	// new certificates and use them to connect.
 	if nodeDetails.MFACheck != nil && nodeDetails.MFACheck.Required {
-		am, err := proxyClient.sessionSSHCertificate(ctx, nodeDetails)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-
-		authMethods = am
+		clt, err := tc.connectToNodeWithMFA(ctx, proxyClient, nodeDetails, user, details)
+		return clt, trace.Wrap(err)
```
Comment (smallinsky) [inline lib/client/api.go:1487]: 👍  It so much cleaner now.

File: lib/client/api.go [hunk 2]
```diff
@@ -1457,78 +1457,127 @@ func (tc *TeleportClient) SSH(ctx context.Context, command []string, runLocally
...
+	defer directCancel()
+	go func() {
+		// try connecting to the node with the certs we already have
+		clt, err := proxyClient.ConnectToNode(directCtx, nodeDetails, user, details, proxyClient.authMethods)
+		directResultC <- clientRes{clt: clt, err: err}
+	}()
+
+	// use a child context so the goroutine ends if this
+	// function returns early
+	mfaCtx, mfaCancel := context.WithCancel(ctx)
```
Comment (smallinsky) [inline lib/client/api.go:1510]: Why do we need to use a separate context for `ConnectToNode` and `connectToNodeWithMFA` calls ?

File: lib/client/api.go [hunk 3]
```diff
@@ -1495,22 +1496,83 @@ func (tc *TeleportClient) ConnectToNode(ctx context.Context, proxyClient *ProxyC
...
+
+			connectErr = res.err
+		case res := <-mfaResultC:
+			if res.clt != nil {
+				return res.clt, nil
+			}
+		}
+	}
+
+	return nil, trace.Wrap(connectErr)
```
[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/client/api.go:1536]: but the `connectToNodeWithMFA` can fails for other reason for instance MFA device failure/misconfiguration. In this case a user will only see the `connectToNodeWithMFA` Access Denied error and we will lost a bit visibility what going on with `connectToNodeWithMFA` flow.

**Review by smallinsky — APPROVED** [review]

**Review by smallinsky — COMMENTED** [review]
Seems that UT is broken: 
```
{"caller":"web/terminal.go:699","component":"websocket","error":"access denied to root connecting to a31f1bb3-4bdb-4d37-b592-9258578d6bd0:0, close tcp 127.0.0.1:60034-\u003e127.0.0.1:37117: use of closed network connection","level":"warning","message":"Unable to stream terminal - failure connecting to host","session_id":"89ef396c-f46b-4d49-874c-ab2a54720f8c","timestamp":"2023-03-31T13:44:13Z"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x58ba51e]

goroutine 31640 [running]:
github.com/gravitational/teleport/lib/client.(*NodeClient).RunInteractiveShell(0x0, {0x93c90d8, 0xc01dd3a330}, {0x7af6eaf, 0x4}, {0x0, 0x0})
	/__w/teleport/teleport/lib/client/client.go:1604 +0x11e
github.com/gravitational/teleport/lib/web.(*TerminalHandler).streamTerminal(0xc0099338c0, 0xc003ed7520?, 0xc0013c7860?)
	/__w/teleport/teleport/lib/web/terminal.go:705 +0x42b
created by github.com/gravitational/teleport/lib/web.(*TerminalHandler).handler
	/__w/teleport/teleport/lib/web/terminal.go:425 +0xbfa
FAIL	github.com/gravitational/teleport/lib/web	144.483s
===================================================
Makefile:617: recipe for target 'test-go-unit' failed
```

File: lib/client/api.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/client/api.go]: @rosstimothy 
It looks that we are still returning either error from MFA flow or direct call. 

The `MFARequiredUnknownErr`  allow only to distinguish the AccessDenied error  that is returned from the API from MFA ceremony. 

can we just add something like: 

```go
if trace.IsAccessDenied(mfaFlowErr) || trace.IsAccessDenied(noMFAFlowErr)  {
   return trace.AccessDenied(trace.Aggragate(mfaFlowErr, noMFAFlowErr).Error())
}

[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/client/api.go]: That makes sense. Then I think that solution with  `MFARequiredUnknownErr` is suitable but can we wrap the MFARequiredUnknownErr with `trace.Wrap(NewMFARequire())` to preserve stack trace ?

**Review by atburke — APPROVED** [review]

File: lib/service/service.go [hunk 4]
```diff
@@ -3849,6 +3849,7 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 	}
 
 	tlscfg := serverTLSConfig.Clone()
+	setupTLSConfigClientCAsForCluster(tlscfg, accessPoint, clusterName)
```
Comment (greedy52) [inline lib/service/service.go:3852]: probably not a big deal but `setupTLSConfigClientCAsForCluster` sets `tlsClone.ClientAuth = tls.VerifyClientCertIfGiven` which conflict with next line.

**Pull request merged**